agent-protocol.md

```diff
@@ -0,0 +1,224 @@
+- [SMP agent commands](#smp-agent-commands)
+- [Connection invitation](#connection-invitation)
+
+## Abstract
+
+The purpose of SMP agent protocol is to define the syntax and the semantics of communications between the client and the agent that connects to [SMP](./simplex-messaging.md) servers.
+
+It provides:
+- convenient protocol to create and manage bi-directional (duplex) connection to the users of SMP agents consisting of two separate unidirectional (simplex) SMP queues, abstracting away multiple steps required to establish bi-directional connections.
+- management of E2E encryption between SMP agents, generating ephemeral RSA keys for each connection.
+- SMP command authentication on SMP servers, generating ephemeral RSA keys for each SMP queue.
+- TCP transport handshake and encryption with SMP servers.
+- validation of message integrity.
+
+SMP agent protocols provides no encryption or any security on the client side - it is assumed that the agent is executed in the trusted and secure environment.
+
+The future versions of this protocol could provide:
+- managing redundant SMP queues with more than 1 queue in each direction.
+- managing simple symmetric groups as a foundation for chat groups and device synchronization.
+- agent cluster - synchronizing states of multiple agents.
+- secure "synchronous" streams with symmetric message encryption and connection-level authentication (requires extending SMP protocol) - it can be used, e.g., for file transfers.
+
+## SMP Agent
+
+SMP agent is a client-side process or library that communicates via SMP servers using [simplex messaging protocol (SMP)](./simplex-messaging.md) with other SMP agents according to the commands received from its users. This protocol is a middle layer in SMP protocols stack (above SMP protocol but below any application level protocol) - it is intended to be used by client-side applications that need secure asynchronous bi-directional communication channels ("connections").
+
+The agent must have a persistent storage to manage the states of known "connections" and of the client-side information of two SMP queues that each connection consists of, and also the buffer of the most recent messages. The number of the messages that should be stored is implementation specific, depending on the error management approach that the agent implements; at the very least the agent must store the hash and id of the last received message.
+
+## SMP agent protocol components
+
+SMP agent protocol has 3 main parts:
+
+- the syntax and semantics of messages that SMP agents exchange between each other to negotiate establishing unidirectional (simplex) encrypted queues on SMP server(s) to provide their users a higher level interface than SMP protocol to establish and operate duplex (bi-directional) connections and notifications about any communication integrity violations (via hash validation, sequential message IDs and message acknowledgements send over SMP queues).
```

Comment:
Kind of hard to digest this sentence, split it up (maybe a couple times, but at least at the point "and notifications")? Took me several tries to read :) And I'm still struggling to understand what refers to notifications - "establish and operate notifications"..? "syntax and semantics of notifications"? These can't be right

Reply:
or maybe it's "provide users notifications" :)